### PR TITLE
Add model architecture for minGRU eval on SQuAD

### PR DESCRIPTION
**Pull request by dlrudwo1269**
- Add minGRU architecture adapted for SQuAD
- Add data loading and preprocessing function
- Add training loop
- Allow masking of padding tokens in minGRU.forward()

I'm pretty sure my logic for handling padding/masking works (from the MinGRUSquadQA model level down to the minGRU layer level, but another set of eyes on it would be very appreciated)

Ran a couple epochs of training on a subset of the training data:
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 743/743 [02:38<00:00,  4.68it/s]
Epoch    0 | loss: 0.792170 | cur_lr: 0.000015 | dt: 158622.10 | tok/sec: 17242.16
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 743/743 [02:23<00:00,  5.17it/s]
Epoch    1 | loss: 0.583884 | cur_lr: 0.000030 | dt: 143616.72 | tok/sec: 19043.66

### CODE REVIEW

File: src/layers/MinGRU.py [hunk 1]
```diff
@@ -72,4 +76,10 @@ def forward(self, x, h_prev=None):
             log_tilde_h = self.log_g(tilde_h) # Log candidate state
             h = self.parallel_scan_log(log_one_minus_z, log_z + log_tilde_h) # Hidden states
 
+            if mask is not None:
+                mask = mask.unsqueeze(-1) # [batch_size, seq_len, 1]
+                # NOTE: because minGRU computes hidden states sequentially (unlike attention)
+                #       if suffices to mask out the irrelevant hidden states to 0
+                h = h.masked_fill(mask, 0)
+
```
Comment (lbaierreinio) [inline src/layers/MinGRU.py:84]: This looks good to me. I'll just need to update the synthetic dataset to add that [CLS] token back in and update the classifier to find the [CLS] token rather than use the last token in the sequence for prediction

File: train_squad.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            y = torch.cat([answer_start_idx, answer_end_idx], dim=-1) # [batch_size, 2]
+            mask = (x == tokenizer.pad_token_id)
+            if ampere_gpu:
+                # mixed precision training
+                with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
+                    logits, loss = model(x, targets=y, mask=mask)
+            else:
+                logits, loss = model(x, targets=y, mask=mask)
+            val_loss_accum += loss.detach()
+
```
Comment (lbaierreinio) [inline train_squad.py:135]: Might be missing something but I'm pretty sure we'd want to extract this into a function rather than duplicating code

**Review by lbaierreinio — APPROVED** [review]

File: src/layers/MinGRU.py (hunk 1 above)
Comment (dlrudwo1269) [reply src/layers/MinGRU.py:84]: Will address in https://github.com/lbaierreinio/min-gru-transformers/issues/27
